src/reduced_order/pod_basis_offline.cpp

```diff
@@ -0,0 +1,156 @@
+
+    for (const auto & entry : files_in_directory){
+        if(std::string(entry.filename()).std::string::find("solution_snapshot") != std::string::npos){
+            pcout << "Processing " << entry << std::endl;
+            file_found = true;
+            std::ifstream myfile(entry);
+            if(!myfile)
+            {
+                pcout << "Error opening file." << std::endl;
+            }
+            std::string line;
+            int rows = 0;
+            int cols = 0;
+            //First loop set to count rows and columns
+            while(std::getline(myfile, line)){ //for each line
+                std::istringstream stream(line);
+                std::string field;
+                cols = 0;
+                while (getline(stream, field,' ')){ //parse data values on each line
+                    if (field.empty()){ //due to whitespace
+                        continue;
+                    } else {
+                        cols++;
+                    }
+                }
+                rows++;
+            }
+
+            snapshotMatrix.conservativeResize(rows, snapshotMatrix.cols()+cols); //Subtract 1 from row because of header row
+
+            int row = 0;
+            myfile.clear();
+            myfile.seekg(0); //Bring back to beginning of file
+            //Second loop set to build solutions matrix
+            while(std::getline(myfile, line)){ //for each line
+                std::istringstream stream(line);
+                std::string field;
+                int col = 0;
+                while (getline(stream, field,' ')) { //parse data values on each line
+                    if (field.empty()) {
+                        continue;
+                    } else {
+                        snapshotMatrix(row, snapshotMatrix.cols()-cols+col) = std::stod(field); //This will work for however many solutions in each file
+                        col++;
+                    }
+                }
+                row++;
+            }
+            myfile.close();
+        }
+    }
+
+    pcout << "Snapshot matrix generated." << std::endl;
+
+
+    /* Reference for simple POD basis computation: Refer to Algorithm 1 in the following reference:
+    "Efficient non-linear model reduction via a least-squares Petrov–Galerkin projection and compressive tensor approximations"
+    Kevin Carlberg, Charbel Bou-Mosleh, Charbel Farhat
+    International Journal for Numerical Methods in Engineering, 2011
+    */
+
+    pcout << "Computing POD basis..." << std::endl;
+
+    VectorXd reference_state = snapshotMatrix.rowwise().mean();
+
+    referenceState.reinit(reference_state.size());
+    for(unsigned int i = 0 ; i < reference_state.size() ; i++){
+        referenceState(i) = reference_state(i);
+    }
+
+    snapshotMatrix = snapshotMatrix.colwise() - reference_state;
+
+    Eigen::BDCSVD<MatrixXd, Eigen::DecompositionOptions::ComputeThinU> svd(snapshotMatrix);
+    MatrixXd pod_basis = svd.matrixU();
```

Comment:
That does the SVD decomposition? Eigen seems quite convenient!

Reply:
Very convenient, I highly recommend it!